Source/WebCore/platform/adwaita/ThemeAdwaita.cpp

```diff
@@ -51,46 +49,34 @@ Theme& Theme::singleton()
 
 ThemeAdwaita::ThemeAdwaita()
 {
-#if PLATFORM(GTK)
-    if (auto* settings = gtk_settings_get_default()) {
-        refreshGtkSettings();
-
-        // Note that Theme is NeverDestroy'd so the destructor will never be called to disconnect this.
-        g_signal_connect_swapped(G_OBJECT(settings), "notify::gtk-enable-animations", G_CALLBACK(+[](ThemeAdwaita* theme, GParamSpec*, GObject*) {
-            theme->refreshGtkSettings();
-        }), this);
-#if !USE(GTK4)
-        g_signal_connect_swapped(G_OBJECT(settings), "notify::gtk-theme-name", G_CALLBACK(+[](ThemeAdwaita* theme, GParamSpec*, GObject*) {
-            theme->refreshGtkSettings();
-        }), this);
-#endif // !USE(GTK4)
-    }
-
-#endif // PLATFORM(GTK)
+#if PLATFORM(GTK) || PLATFORM(WPE)
+    refreshSettings();
+
+    // Note that Theme is NeverDestroy'd so the destructor will never be called to disconnect this.
+    SystemSettingsGLib::singleton().addObserver([this](const SettingsStateGLib &state) mutable {
+        if (state.enableAnimations || state.themeName)
+            this->refreshSettings();
+    }, this);
+#endif // PLATFORM(GTK) || PLATFORM(WPE)
 }
 
-#if PLATFORM(GTK)
+#if PLATFORM(GTK) || PLATFORM(WPE)
 
-void ThemeAdwaita::refreshGtkSettings()
+void ThemeAdwaita::refreshSettings()
 {
-    if (auto* settings = gtk_settings_get_default()) {
-        gboolean enableAnimations;
-        g_object_get(settings, "gtk-enable-animations", &enableAnimations, nullptr);
-        m_prefersReducedMotion = !enableAnimations;
-
-        // For high contrast in GTK3 we can rely on the theme name and be accurate most of the time.
-        // However whether or not high-contrast is enabled is also stored in GSettings/xdg-desktop-portal.
-        // We could rely on libadwaita, dynamically, to re-use its logic but, no matter how we do it, the setting
-        // has to be proxied over from the UI process different than how GtkSettings is.
+    if (auto enableAnimations = SystemSettingsGLib::singleton().enableAnimations())
+        m_prefersReducedMotion = !enableAnimations.value();
+
+    // For high contrast in GTK3 we can rely on the theme name and be accurate most of the time.
+    // However whether or not high-contrast is enabled is also stored in GSettings/xdg-desktop-portal.
+    // We could rely on libadwaita, dynamically, to re-use its logic.
 #if !USE(GTK4)
-        GUniqueOutPtr<char> gtkThemeName;
-        g_object_get(settings, "gtk-theme-name", &gtkThemeName.outPtr(), nullptr);
-        m_prefersContrast = !g_strcmp0(gtkThemeName.get(), "HighContrast") || !g_strcmp0(gtkThemeName.get(), "HighContrastInverse");
+    if (auto themeName = SystemSettingsGLib::singleton().themeName())
+        m_prefersContrast = themeName == "HighContrast"_s || themeName == "HighContrastInverse"_s;
 #endif // !USE(GTK4)
```

Comment:
If this is specific to GTK3 and this method is also called for WPE now, we should probably use `#if PLATFORM(GTK) && !USE(GTK4)`